core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1917,6 +1917,27 @@ public RexLiteral makeZeroLiteral(RelDataType type) {
     return makeLiteral(zeroValue(type), type);
   }
 
+  public RexNode makeZeroForNestedType(RelDataType type) {
+    SqlOperator op;
+    switch (type.getSqlTypeName()) {
+    case ARRAY:
+      op = SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR;
+      break;
+    case MULTISET:
+      op = SqlStdOperatorTable.MULTISET_VALUE;
+      break;
+    case MAP:
+      op = SqlStdOperatorTable.MAP_VALUE_CONSTRUCTOR;
+      break;
+    case ROW:
+      op = SqlStdOperatorTable.ROW;
```

Comment:
I don't think this works for ROW. For ROW your original solution was correct: you need to create a zero for each field using a recursive call.

Reply:
I tried to sort out the logic according to your suggestion. ROW needs to be processed recursively and filled with real zero values. Should MAP have the same logic as ROW or ARRAY? If both use recursive logic, then ARRAY and MULITSET only need to do a CAST at the innermost level?

Reply:
Yes, ROW needs real zeros, MAP, ARRAY, and MULTISET should be empty.
The empty value is the zero value for monoid provided by the concatenation operation for these data structures.

Reply:
Done!